amptorch/descriptor/MCSH/__init__.py

```diff
@@ -171,55 +174,76 @@ def calculate_fingerprints(self, atoms, element, calc_derivatives, log):
-                    atom_indices_p, atom_num, cal_atoms_p, cal_num, \
-                    self.params_set['ip'], self.params_set['dp'], self.params_set['num'], self.params_set['gaussian_params_p'], self.params_set['ngaussians_p'], self.params_set['element_index_to_order_p'],\
-                    x_p, dx_p)
-        
-        if errno == 1:
-            raise NotImplementedError("Descriptor not implemented!")
-        fp = np.array(x)
-        fp_prime = np.array(dx)
-
-        # if "prime_threshold" in self.params_set:
-        #     threshold = self.params_set["prime_threshold"]
-        #     super_threshold_indices = np.abs(fp_prime) < threshold
-        #     print("threshhold: {} \tnum points set to zero:{} \t outof: {}".format(threshold, np.sum(super_threshold_indices), fp_prime.shape[0] * fp_prime.shape[1]))
-        #     fp_prime[super_threshold_indices] = 0.0
-
-        scipy_sparse_fp_prime = sparse.coo_matrix(fp_prime)
-        # print(fp)
-        # print(fp.shape)
-        # print(np.sum(super_threshold_indices))
-        # print(np.min(np.abs(scipy_sparse_fp_prime.data)))
-        # print("density: {}% \n\n----------------------".format(100*len(scipy_sparse_fp_prime.data) / (fp_prime.shape[0] * fp_prime.shape[1])))
-
-        return (
-                size_info,
-                fp,
-                scipy_sparse_fp_prime.data,
-                scipy_sparse_fp_prime.row,
-                scipy_sparse_fp_prime.col,
-                np.array(fp_prime.shape),
-            )
+        if calc_derivatives:
+            x = np.zeros([cal_num, self.params_set['num']], dtype=np.float64, order='C')
+            dx = np.zeros([cal_num * self.params_set['num'], atom_num * 3], dtype=np.float64, order='C')
+
+            x_p = _gen_2Darray_for_ffi(x, ffi)
+            dx_p = _gen_2Darray_for_ffi(dx, ffi)
+
+            if self.params_set['square']:
+                errno = lib.calculate_atomistic_mcsh_square(cell_p, cart_p, scale_p, pbc_p,\
+                            atom_indices_p, atom_num, cal_atoms_p, cal_num, \
+                            self.params_set['ip'], self.params_set['dp'], self.params_set['num'], self.params_set['gaussian_params_p'], self.params_set['ngaussians_p'], self.params_set['element_index_to_order_p'],\
+                            x_p, dx_p)
+            else:
+                errno = lib.calculate_atomistic_mcsh(cell_p, cart_p, scale_p, pbc_p,\
+                            atom_indices_p, atom_num, cal_atoms_p, cal_num, \
+                            self.params_set['ip'], self.params_set['dp'], self.params_set['num'], self.params_set['gaussian_params_p'], self.params_set['ngaussians_p'], self.params_set['element_index_to_order_p'],\
+                            x_p, dx_p)
+            
+            if errno == 1:
+                raise NotImplementedError("Descriptor not implemented!")
+            fp = np.array(x)
+            fp_prime = np.array(dx)
+
+            # if "prime_threshold" in self.params_set:
```

Comment:
do we need these, can you clean up these comments before merge?

Reply:
yes... I'm experimenting how to make MCSH more robust with different ideas, these will be gone once there is a set protocal
For now I would love to keep it there for developing/debugging purpose if it's ok